### PR TITLE
Enable disable restart get and set snap operations

### PR DESCRIPTION
**Pull request by rmescandon**
Additional snap operations of help in some cases

### CODE REVIEW

File: lib/charms/layer/snap.py [hunk 1]
```diff
@@ -105,6 +105,61 @@ def connect_all():
             connect(plug, slot)
 
 
+def disable(snapname):
+    '''Disables a snap in the system
+
```
Comment (stub42) [inline lib/charms/layer/snap.py:110]: Also on a separate line:

Sets the snap.disabled.{snapname} flag

File: lib/charms/layer/snap.py [hunk 2]
```diff
@@ -105,6 +105,61 @@ def connect_all():
...
+    This method doesn't affect any snap state if requested snap does not
+    exist
+    '''
+    hookenv.log('Disabling {} snap'.format(snapname))
+    subprocess.check_call(['snap', 'disable', snapname],
+                          universal_newlines=True)
+    reactive.set_state('snap.disabled.{}'.format(snapname))
```
Comment (stub42) [inline lib/charms/layer/snap.py]: set_state is now deprecated in favor of set_flag, but its fine to leave it as you have it; I will do a complete pass through the layer at some point.

File: lib/charms/layer/snap.py [hunk 3]
```diff
@@ -105,6 +105,61 @@ def connect_all():
...
+
+
+def enable(snapname):
+    '''Enables a snap in the system
+
```
Comment (stub42) [inline lib/charms/layer/snap.py:130]: Also add a separate line:

Clears the snap.disabled.{snapname} flag.

File: lib/charms/layer/snap.py [hunk 4]
```diff
@@ -105,6 +105,61 @@ def connect_all():
...
+    '''
+    hookenv.log('Set config {}={} for snap {}'.format(key, value, snapname))
+    subprocess.check_call(
+        ['snap', 'set', snapname, '{}={}'.format(key, value)])
+
+
+def get(snapname, key):
+    '''Gets configuration options for a snap
+
+    This method returns the output that snactl get command prints out.
```
Comment (stub42) [inline lib/charms/layer/snap.py]: typo in snactl

File: lib/charms/layer/snap.py [hunk 5]
```diff
@@ -105,6 +105,61 @@ def connect_all():
...
+
+
+def set(snapname, key, value):
+    '''Changes configuration options in a snap
+
+    This method will fail if snapname is not an installed snap
+    '''
+    hookenv.log('Set config {}={} for snap {}'.format(key, value, snapname))
+    subprocess.check_call(
+        ['snap', 'set', snapname, '{}={}'.format(key, value)])
```
Comment (stub42) [inline lib/charms/layer/snap.py]: I think we are fine without validation here. No shell escaping is needed. A key containing an = might be an issue, but I can't really see that being user input in any case where the value is not also user input.

**Review by stub42 — CHANGES_REQUESTED** [review]
I'm nervous about relying on the 'snap' command returning a success state for operations on non-existent snaps. That seems like a bug in 'snap' to me, as it makes automation impossible without race conditions. I think we should guard the operations with 'if reactive.is_flag_set('snap.installed.{}'.format(snapname)), so we only attempt operations on snaps that were added with the snap layer API. We should log a warning if the flag was not set and we chose to do nothing.

It does mean that charms need to explicitly list snaps that come preinstalled for installation, but I think this is a good thing as charms should not make assumptions about what comes preinstalled; those assumptions can change.

File: lib/charms/layer/snap.py (hunk 1 above)
Comment (rmescandon) [reply lib/charms/layer/snap.py:110]: Set that comment and also the analog one in enable() method

File: lib/charms/layer/snap.py (hunk 2 above)
Comment (rmescandon) [reply lib/charms/layer/snap.py]: Changed here and in enable() method

File: lib/charms/layer/snap.py (hunk 4 above)
Comment (rmescandon) [reply lib/charms/layer/snap.py]: Fixed

File: lib/charms/layer/snap.py (hunk 5 above)
Comment (rmescandon) [reply lib/charms/layer/snap.py]: Nothing changed here. Is that fine or you want to add some extra bits?

File: lib/charms/layer/snap.py [hunk 6]
```diff
@@ -105,6 +105,85 @@ def connect_all():
...
+    Sets the snap.disabled.{snapname} flag
+
+    This method doesn't affect any snap state if requested snap does not
+    exist
+    '''
+    hookenv.log('Disabling {} snap'.format(snapname))
+    if not reactive.is_flag_set('snap.installed.{}'.format(snapname)):
+        hookenv.log(
+            'Cannot disable {} snap because it is not installed'.format(
+                snapname), hookenv.WARNING)
```
Comment (stub42) [inline lib/charms/layer/snap.py:120]: Needs a 'return' after the log message. Here and the others.

Comment (rmescandon) [reply lib/charms/layer/snap.py:120]: ouch!. Done

**Pull request merged**